common/src/main/templates/io/netty/util/collection/KObjectHashMap.template

```diff
@@ -350,9 +350,17 @@ private int indexOf(int key) {
     /**
      * Returns the hashed index for the given key.
      */
-    private int hashIndex(int key) {
+    private int hashIndex(@k@ key) {
         // Allowing for negative keys by adding the length after the first mod operation.
-        return (key % keys.length + keys.length) % keys.length;
+        int hashCode = hashCode(key);
+        return (hashCode % keys.length + keys.length) % keys.length;
+    }
+
+    /**
+     * Returns the hash code for the key.
+     */
+    private int hashCode(@k@ key) {
+       return @O@.valueOf(key).hashCode();
```

Comment:
Unfortunately the static `hashCode(...)` values don't appear until Java 8 ... *sigh*. I've modified our template to generate the correct hash code based on the primitive type.